src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -187,17 +233,19 @@ def _backchain_one_pass(
 
                     # After all this, the unification call that failed earlier
                     # (leading us into the current else statement) should work.
-                    ground_op = self._find_unification(necessary_add_effects,
-                                                       pnad, segment)
+                    best_score_pnad, ground_op = self._find_unification(
+                        necessary_add_effects,
+                        param_opt_to_nec_pnads[option.parent], segment)
                     assert ground_op is not None
+                    assert best_score_pnad == pnad
                     obj_to_var = dict(
                         zip(ground_op.objects, pnad.op.parameters))
 
                 # Every atom in the necessary_image that wasn't in the
-                # necessary_add_effects is a possible keep effect. This
+                # ground_op's add effects is a possible keep effect. This
                 # may add new variables, whose mappings for this segment
                 # we keep track of in the seg_to_keep_effects_sub dict.
-                for atom in necessary_image - necessary_add_effects:
+                for atom in necessary_image - ground_op.add_effects:
```

Comment:
hmm, so this is a functionality change. was it wrong before? or did you just find this to work better?